Experiments with Python interface

timmoon10 opened the pull request:
Consider this a series of design experiments, all of which are open to debate:
- There is now a `Layer` class that is distinct from `Module`. A `Layer` is now like a single PyTorch operation (takes an input and produces an output) and a `Module` is now like a PyTorch module (can be applied repeatedly to different inputs). I personally think LBANN uses the word "layer" in a somewhat non-standard way and it should be changed to something like "operation", but that's a question for another day.
- You can specify a `Layer`'s input in the constructor. It's now possible to specify a network in one pass:
```python
conv1 = lp.Convolution('conv1', x, other params)
bn1 = lp.BatchNormalization('bn1', conv1, other params)
relu1 = lp.Relu('relu1', bn1', other params)
```
- The data layout is now an optional parameter when constructing a `Layer`. The default is data-parallel.
- I prefer pure functions over functions with side effects (something something functional programming). To export a `Layer` to prototext, you call `Layer.export_proto()` to get a protobuf layer message and then you are responsible for adding it to the protobuf model message. I've also reworked the layer subclass generator to use a pure function.
- I've reworked the `traverse_model` function, which applies a ~DFS on the layer graph and applies a function to each layer. Now, the `traverse_layer_graph` function (models are more than just their layers) returns a list of layers and you can apply `map` afterwards. In the future, I think it would be best if `traverse_layer_graph` did a DFS on the graph (finding parent nodes in addition to child nodes) and then did a topological sort.
- I removed the base `Module` class since it has nothing to implement, but I don't think this is the right thing to do. It would be better for `Module` to be an abstract base class.
- I've moved ResNet-specific modules within resnet50.py rather than in a globally accessible module, which is closer to PyTorch's approach (https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py).
- The "parent"/"child" layer terminology has been replaced with "input"/"output". I think this is more natural, but perhaps we want to stick with the terminology we use in the prototext.
- I have a fondness for the conditional operator.

Commit:
Merge branch 'develop' into pyproto

Commit:
Merge branch 'pyproto' of https://github.com/ndryden/lbann into pyproto

Commit:
Initial tinkering with Python interface.

Distinguishing layers and modules. Preferring pure functions. Moving ResNet-specific modules to ResNet script.

Commit:
Merge branch 'develop' into pyproto

Commit:
Tinkering until Python script for ResNet-50 works.

Weights are not properly setup.

Commit:
Merge branch 'pyproto' of https://github.com/ndryden/lbann into pyproto

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -2,6 +2,7 @@
 
 import sys
 import google.protobuf.text_format
+import collections
```
Should be `import collections.abc` (see below).

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
-        w = Weights(self.name + name_suffix, initializer)
+        # Initialize inputs, outputs, and weights
+        for l in (inputs
+                  if isinstance(inputs, collections.Iterable)
```
The check should be `isinstance(inputs, collections.abc.Iterable)`. Using the abstract base classes from `collections` is deprecated (and will be removed in Python 3.8).

timmoon10 replied:
Done.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
+                  else [inputs]):
+            self.add_input(l)
+        for l in (outputs
+                  if isinstance(outputs, collections.Iterable)
```
`collections.abc`

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
+                  else [outputs]):
+            self.add_output(output)
+        for w in (weights
+                  if isinstance(weights, collections.Iterable)
```
`collections.abc`

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
-        self.parents = []
-        self.children = []
+    def __init__(self, name='', inputs=[], outputs=[],
+                 weights=[], data_layout=''):
```
I like making the data layout optional, but while I'm aware LBANN will default it to `data_parallel` if it's not present, I think it's worth being explicit about it when writing to the prototext. So I would just change the default to `data_layout='data_parallel'`.

timmoon10 replied:
Done.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -371,34 +372,37 @@ def proto(self, model):
 
-def traverse_model(first_module, func):
-    """Traverse a model and apply func to each layer.
+def traverse_layer_graph(start_layers):
```
Going this route, I think it would be better to make this a generator instead of explicitly returning a list. It's more Pythonic.

timmoon10 replied:
Done. Our traversals may be a bit heavy-weight (DFS to find root nodes, DFS for topologically sorted traversal), but I think safety is more important than performance here.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -371,34 +372,37 @@ def proto(self, model):
+    return layers
 
-def save_model(filename, first_module, data_layout, mini_batch_size, epochs,
+def save_model(filename, layers, mini_batch_size, epochs,
```
1. Let's keep the data layout argument, just make it default to `data_parallel` (as above).
2. I think `first_module` (or `first_layer`, or maybe `first_modules`/`layers` since we could have multiple input layers, I guess) is a better argument name than `layers`, which reads like you need to provide _all_ the layers in the network.

timmoon10 replied:
1. As far as I'm aware of, the `data_layout` field in the `Model` protobuf message doesn't do anything.
2. I think it might be simpler if we expected the user to provide all the layers. It's fairly straightforward with the `traverse_layer_graph` function.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -410,7 +414,8 @@ def save_model(filename, first_module, data_layout, mini_batch_size, epochs,
         callback.proto(pb.model)
-    traverse_model(first_module, lambda l: l.proto(pb.model))
+    layers = traverse_layer_graph(layers)
+    pb.model.layer.extend([l.export_proto() for l in layers])
```
If `traverse_layer_graph` becomes a generator, this can turn into `pb.model.layer.extend([l.export_proto() for l in traverse_layer_graph(layers)])`.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
         self.weights = []
+        self.data_layout = data_layout
 
-    def __call__(self, parent):
```
I think we should keep the callable syntax for chaining layers together.

timmoon10 replied:
Done.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -41,85 +42,78 @@ def _add_to_module_namespace(stuff):
 
-    def _add_layer(self, model):
-        """Add a layer entry to model (for use by proto).
+def _create_layer_subclass(type_name, layer_field_name):
```
This definitely needed improved documentation, otherwise I'd look at it in a month and it would be black magic. Thanks! 😄

ndryden reviewed (changes requested):
Overall I think this is moving in a good direction.

On input/output versus parent/child: I'm not so concerned about the terminology in the prototext as long as we can keep a clear interface. I tend to think this terminology differences depends on whether you think of this as constructing a graph, or feeding layers into each other. Whichever you think is most clear is fine with me for this.

Broader changes:
* I think it's worth keeping the ResNet building blocks in `blocks.py`. ResNets (and their components) are widely-used enough that I think they can go there. In the future, as we collect more building blocks, we can look at splitting them out.
* Let's reintroduce `Module` as an [abstract base class](https://docs.python.org/3/library/abc.html) (it might be worth having `Layer` inherit from it, since a layer is just a module with a single op). It probably needs abstract `export_proto` and `__call__` methods.
* Let's standardize toward the approach you've used for `Layer`, to keep the rest of the API uniform.

We also need to figure out how to handle weights associated with layers. I have some thoughts I'll send you on Slack.

Commit:
Updating Python interface for weights.

Similar to Python interface for layers.

Commit:
Updating Python interface for objective functions and metrics.

Similar to Python interface for layers.

Commit:
Updating Python interface for callbacks.

Similar to Python interface for layers.

Commit:
Adopting review suggestions from @ndryden for Python interface.

Implemented helper function to convert objects to iterators. Calling Layer object will add an input.

Commit:
Tinkering with Python interface for layers.

Changing "input"/"output" terminology back to "parent"/"child". Making layer names an advanced option.

Commit:
Layer graph traversal in Python interface is topologically ordered.

traverse_layer_graph is now a generator function that traverses all connected layers in layer graph (not just descendants).

Commit:
Restoring neural network modules in Python interface.

Modules are now similar to PyTorch modules/layers. Added fully-connected and convolution modules (similar to FC and conv "layers" in PyTorch/TensorFlow/Keras).

timmoon10 commented:
Ready for another round of review:
- Style changes to `Layer` have been propagated to `Initializer`, `Weights`, `ObjectiveFunctionTerm`, `ObjectiveFunction`, `Metric`, and `Callback`.
- Returned back to "parent"/"child" layer terminology.
- Layer and weight names are now advanced options. Specifying a network looks something like:
```python
conv1 = lp.Convolution(x, other params)
bn1 = lp.BatchNormalization(conv1, other params)
relu1 = lp.Relu(bn1, other params)
```
- `traverse_layer_graph` now works with any start node, not just root nodes. It is a generator function that outputs layers in a topological order. It will probably fail horribly if there is a cycle.
- Restored `Module` and added documentation. It should be used in a similar manner as PyTorch, i.e. a `Module` is a pattern that can be applied repeatedly to different inputs.
- I've documented how our usage of the word "layer" is inconsistent with PyTorch, TensorFlow, and Keras. Their "layers" are similar to our "modules". 
- Implemented fully-connected and convolution modules, which are similar to fully-connected and convolution "layers" in PyTorch/TensorFlow/Keras.

This is ready to merge, ~~although we need to update the readme~~.

Commit:
Merge branch 'pyproto' of https://github.com/ndryden/lbann into pyproto

Commit:
Updating Python interface readme.

Commit:
FC and conv modules will create default weights if none are provided.

ndryden commented on scripts/proto/README.md:
```diff
@@ -1,166 +1,186 @@
+different sets of input tensors, each time obtaining a different set
+of output tensors. Their usage of the word "layer" is closer to our
+usage of the word "module." In fact, PyTorch's "layers" are subclasses
+of their `Module` class.
```
I would clarify that part of this is because, in this Python interface, we're not really doing forward/back prop, but just setting up the structure of the network.

ndryden commented on scripts/proto/README.md:
```diff
@@ -1,166 +1,186 @@
+API.
 
-## Building blocks
+## `lbann_module`
```
Should be `lbann_modules`.

ndryden commented on scripts/proto/lbann_modules.py:
```diff
@@ -0,0 +1,221 @@
+
+        # Initialize activation layer
+        self.activation = None
+        if activation:
```
This prevents using modules as an action (if, for whatever reason, that makes sense). Not sure if we care.

timmoon10 replied:
Thinking about it, a module is actually more natural here. We are basically mimicking a module that applies an activation function. Perhaps we could implement a module that wraps around a single layer? If the user provides a layer, we can construct the corresponding module. Or maybe it would be messier than it's worth.

ndryden replied:
That might be more complex than it's worth. Alternately, we could make layers a subclass of modules. This currently seems like an edge-case we can address later though.

ndryden commented on scripts/proto/lbann_modules.py:
```diff
@@ -0,0 +1,221 @@
+                lp.Weights(initializer=lp.ConstantInitializer(value=0.0)))
+
+        # Initialize activation layer
+        self.activation = None
```
Ditto on modules.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -349,68 +475,73 @@ def __init__(self, **kwargs):
-    The model is traversed in a depth-first order except that no layer is
-    visited until all its ancestors have been visited.
+def save_model(filename, mini_batch_size, epochs,
+               layers=[], weights=[], objective_function=None,
```
Does it ever make sense to pass no layers? If not, maybe we shouldn't default this to `[]`.

timmoon10 replied:
I could imagine someone doing something with just weights and objective functions, but that would be a very weird use-case.

ndryden commented on scripts/proto/lbann_proto.py:
```diff
@@ -349,68 +475,73 @@ def __init__(self, **kwargs):
-    traverse_model(first_module, lambda l: l.proto(pb.model))
+
+    # Add layers
+    layers = list(traverse_layer_graph(layers))
```
If we're calling `traverse_layer_graph` here, do we need to pass all the layers in for `layers`? Should document appropriately.

timmoon10 replied:
This is just for safety. It's an extra graph traversal, but I assume that that is relatively cheap.

ndryden commented on scripts/proto/resnet50.py:
```diff
@@ -1,69 +1,139 @@
+ce = lp.CrossEntropy([softmax, labels])
+top1 = lp.CategoricalAccuracy([softmax, labels])
+top5 = lp.TopKCategoricalAccuracy([softmax, labels], k=5)
+layers = list(lp.traverse_layer_graph(input))
```
`lp.save_model` calls `traverse_layer_graph` internally-- we should do so in one place, not both.

timmoon10 replied:
We need the list of layers here so that we can extract their weights and feed them into L2 regularization. I'd like to call `traverse_layer_graph` inside of `save_model` for safety. Maybe I'm being too cautious, but these graphs should not be egregiously large. If they were really so large that graph traversal were painful, imagine how long it would take for a single forward or backward prop iteration.

ndryden replied:
Fair enough. 👍

ndryden reviewed (commented):
I think this looks good, just minor changes. I will test and approve later tonight.

Vaguely related comment: If this becomes the primary approach for creating prototexts, and most people don't explicitly name their layers, we may want to tone down the amount of output related to layers/weights that LBANN produces by default. It's already very extensive for reasonably-sized models (e.g. ResNet-50), and with automatically-generated names is probably incomprehensible.

timmoon10 commented:
I agree it would be worthwhile to make the layer/weights printing an advanced feature, either a command-line flag or a callback.

Commit:
Small review suggestions from @ndryden.

ndryden commented:
Tested on Lassen, everything seems good for the generated ResNet-50 model.
